### PR TITLE
Created flat tokens as additional output

### PR DESCRIPTION
**Pull request by douweschulte**
Here I created a flat simple token based output, this can be used for syntax colouring and better change diffs. In some cases the output is not very nice yet, especially generics. And sometimes the ID comes peeking out, in structs in enum variants. The code organisation/API is of course open for discussion.

### CODE REVIEW

File: Cargo.toml
Comment (Enselic) [inline Cargo.toml:17]: Would you say it is too early for me to look into this draft PR in some detail? One comment I have after a quick skim-through is that these deps are not needed any longer, right? Sorry if this is a silly comment to make on a draft :)

Comment (douweschulte) [reply Cargo.toml:17]: Yes those are not needed. I will make sure to run clippy all over the PR code to catch things like this.

Comment (Enselic) [reply Cargo.toml:17]: 👍 

In case you didn't find it yet, there is a convenient script you can run locally that simulates CI: `./scripts/run-ci-locally.sh`

**Comment by douweschulte** [comment]
I would say it is fine to look over the code, especially if you find major architectural improvement points it is nice to know that now. Otherwise I would say that the flat tokens (`tokens::Token` and `tokens::TokenStream`) are pretty much final. The rendering code need to support more input and I am thinking of a refactor into something like the `D` you made, instead of a bunch of functions. But the diff token output is pretty unstable still, I am still changes major things in it.

**Comment by douweschulte** [comment]
I would say the rendering into tokens code is pretty much done. It now supports all exposed types from `rustdoc_types::ItemEnum` I tried to make something useful for all, but if you find something can be expanded/changed please tell me. I struggled a bit with the api members I could not easily test (like `Impl`, `ProcMacro` and some others) so if you have some tests which expose those that would be really helpful. Otherwise it would maybe be wise to create some test crate which exposes all types of api members?

File: src/item_iterator.rs
Comment (Enselic) [inline src/item_iterator.rs:192]: Before I dive deeper I would like to stand a high-level choice of yours. Shouldn't we replace `prefix`, `path` and `suffix` with `tokens`? Rather than having them all.

File: src/intermediate_public_item.rs
Comment (Enselic) [inline src/intermediate_public_item.rs:108]: Related to my other comment: Instead of "duplicating" this function like this, shouldn't we make e.g. `IntermediatePublicItem::type_string_for_item()` and `IntermediatePublicItem::prefix()` produce `Token`s instead of strings?

**Review by Enselic — COMMENTED** [review]
After getting more acquainted with your code, I've got the following high-level design question that I would like to understand before I dive deeper into the code.

**Comment by Enselic** [comment]
> Otherwise it would maybe be wise to create some test crate which exposes all types of api members?

That sounds like exactly what the new `comprehensive_api` in-repo test crate is supposed to accomplish. Run this after doing a git fetch of the latest code to see if that is the case:
```
cargo doc --manifest-path ./tests/crates/comprehensive_api/Cargo.toml --open
```

I plan on replacing `thiserror` with another in-repo test crate that has a `proc-macro`, because those apparently must live in a special crate. But apart from that the API in that crate should be pretty broad.

File: src/item_iterator.rs
Comment (Enselic) [reply src/item_iterator.rs:192]: Ok, keeping `path` separate sounds reasonable now that you say so. So then maybe change the type of `prefix` and `suffix` from current `String` to `TokenStream`? So in  `impl Display for PublicItemInner` we do concatenate a token stream (the prefix), a string array (path), and another token stream (the suffix). I *think* it would be prettier to have also the `path` as a token stream, so that all components that represents a `PublicItem` is of the same type. But it's hard to make a final call without trying it out.

I don't like that the same data is represented in different ways in the same struct. I would like there to be a single source of truth. That generally results in a much simpler system. I hope that makes sense. My apologies in advance if I am completely misinterpreting your code.

I want to be clear that I love the work that you are doing, I hope I don't come across as negative here. I just want to give honest code review comments :)

Comment (douweschulte) [reply src/item_iterator.rs:192]: I do agree on having a single source of truth.
For clarity this is the struct I propose:
```rust
struct PublicItem {
    tokens: TokenStream,
    path: Vec<String>,
}
```
The major downside is that the path is encoded both in the `TokenStream` as `Token`s and in the path as `String`s. But I think this is easier to handle then separate `TokenStream`s as a public api.

And I am really grateful for your feedback, having two eyes on the same code always results in better code. And to be very honest I miss feedback from people who can code at approximately my level a bit in my day to day work.

Comment (Enselic) [reply src/item_iterator.rs:192]: I agree that have a single token stream is an attractive proposition. So if you want to take a shot at using the struct you propose I am fine with that. So now my main concern is https://github.com/Enselic/public_items/pull/53#discussion_r841902208, i.e. that I would like to try to re-use the existing "if this then that" code rather than duplicating it. Which you talked about in:

> The rendering code need to support more input and I am thinking of a refactor into something like the D you made

But my advice would be to try to fit tokens in the existing "D-code" rather than replicating it another time. By no means trivial, but I feel like it would be the best approach.

Comment (douweschulte) [reply src/item_iterator.rs:192]: Great, I will work on some refactoring then and see how 'clean' I can get it. I tried a trait refactoring yesterday which did not really produce what I wanted, as I only need a `Render` trait or similar on like 4 types from the `rustdoc_types`. But I will try and see what I can do.

File: src/diff.rs
Comment (Enselic) [inline src/diff.rs:22]: If you are OK with it, I think it would be good to split this PR up where the first part (this PR) does the change that introduces tokens, and another PR that improves diffing. In other words, if we could remove any changes related to diffing from this PR, it would be good I think.

The reason is that we can have a focused discussion on the details of the core change that introduces tokens, without getting distracted by the diffing parts.

If you really want to do it all in one big PR that is perfectly fine with me too. Since you are the one that writes the code, you are in large parts the one that decides how we should work together to finalize this :)

**Review by Enselic — COMMENTED** [review]
This looks really good on a high level I think! Good job! So now let's move down one level of detail, which means looking at the public API of public_items. Let's do a diff (The feeling of meta when doing this never gets tiresome for me):

<details>
  <summary>Click to expand `cargo public-items --diff-git-checkouts main main-douweschulte`</summary>
  

```
Removed items from the public API
=================================
pub fn public_items::PublicItem::ne(&self, other: &PublicItem) -> bool

Changed items in the public API
===============================
pub fn public_items::PublicItem::cmp(&self, other: &PublicItem) -> $crate::cmp::Ordering
pub fn public_items::PublicItem::cmp(&self, other: &Self) -> std::cmp::Ordering
pub fn public_items::PublicItem::eq(&self, other: &PublicItem) -> bool
pub fn public_items::PublicItem::eq(&self, other: &Self) -> bool
pub fn public_items::PublicItem::partial_cmp(&self, other: &PublicItem) -> $crate::option::Option<$crate::cmp::Ordering>
pub fn public_items::PublicItem::partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering>

Added items to the public API
=============================
pub enum public_items::tokens::ChangedTokenStream
pub enum public_items::tokens::Token
pub enum variant public_items::tokens::ChangedTokenStream::Changed
pub enum variant public_items::tokens::ChangedTokenStream::Same(TokenStream)
pub enum variant public_items::tokens::Token::Function(String)
pub enum variant public_items::tokens::Token::Generic(String)
pub enum variant public_items::tokens::Token::Identifier(String)
pub enum variant public_items::tokens::Token::Keyword(String)
pub enum variant public_items::tokens::Token::Kind(String)
pub enum variant public_items::tokens::Token::Lifetime(String)
pub enum variant public_items::tokens::Token::Primitive(String)
pub enum variant public_items::tokens::Token::Qualifier(String)
pub enum variant public_items::tokens::Token::Self_(String)
pub enum variant public_items::tokens::Token::Symbol(String)
pub enum variant public_items::tokens::Token::Type(String)
pub enum variant public_items::tokens::Token::Whitespace
pub fn public_items::diff::ChangedPublicItem::changed_tokens(&self) -> Vec<ChangedTokenStream>
pub fn public_items::diff::ChangedPublicItem::clone(&self) -> ChangedPublicItem
pub fn public_items::diff::PublicItemsDiff::clone(&self) -> PublicItemsDiff
pub fn public_items::tokens::ChangedTokenStream::clone(&self) -> ChangedTokenStream
pub fn public_items::tokens::ChangedTokenStream::eq(&self, other: &ChangedTokenStream) -> bool
pub fn public_items::tokens::ChangedTokenStream::fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
pub fn public_items::tokens::ChangedTokenStream::ne(&self, other: &ChangedTokenStream) -> bool
pub fn public_items::tokens::Token::clone(&self) -> Token
pub fn public_items::tokens::Token::eq(&self, other: &Token) -> bool
pub fn public_items::tokens::Token::fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
pub fn public_items::tokens::Token::function(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::generic(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::identifier(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::keyword(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::kind(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::len(&self) -> usize
pub fn public_items::tokens::Token::lifetime(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::ne(&self, other: &Token) -> bool
pub fn public_items::tokens::Token::primitive(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::qualifier(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::self_(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::symbol(text: impl Into<String>) -> Self
pub fn public_items::tokens::Token::text(&self) -> &str
pub fn public_items::tokens::Token::type_(text: impl Into<String>) -> Self
pub fn public_items::tokens::TokenStream::clone(&self) -> TokenStream
pub fn public_items::tokens::TokenStream::default() -> TokenStream
pub fn public_items::tokens::TokenStream::eq(&self, other: &TokenStream) -> bool
pub fn public_items::tokens::TokenStream::extend(&mut self, tokens: impl Into<Self>)
pub fn public_items::tokens::TokenStream::fmt(&self, f: &mut $crate::fmt::Formatter<'_>) -> $crate::fmt::Result
pub fn public_items::tokens::TokenStream::fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result
pub fn public_items::tokens::TokenStream::from(token: Token) -> Self
pub fn public_items::tokens::TokenStream::from(tokens: &[Token]) -> Self
pub fn public_items::tokens::TokenStream::from(tokens: Vec<Token>) -> Self
pub fn public_items::tokens::TokenStream::is_empty(&self) -> bool
pub fn public_items::tokens::TokenStream::len(&self) -> usize
pub fn public_items::tokens::TokenStream::ne(&self, other: &TokenStream) -> bool
pub fn public_items::tokens::TokenStream::push(&mut self, token: Token)
pub fn public_items::tokens::TokenStream::remove_from_back(&mut self, len: usize)
pub fn public_items::tokens::TokenStream::tokens(&self) -> impl Iterator<Item = &Token> + '_
pub fn public_items::tokens::TokenStream::tokens_len(&self) -> usize
pub macro public_items::ws!
pub mod public_items::tokens
pub struct public_items::tokens::TokenStream
pub struct field public_items::PublicItem::tokens: tokens::TokenStream
pub struct field public_items::tokens::ChangedTokenStream::Changed::inserted: TokenStream
pub struct field public_items::tokens::ChangedTokenStream::Changed::removed: TokenStream
pub struct field public_items::tokens::TokenStream::tokens: Vec<Token>
```

</details>

Disregarding diffing (see separate comment about that), one high-level comment I have is that I think it would be good if we could remove all API that allows clients to create tokens themselves. Having such an API vastly complicates upholding backwards compatibility. I think it is good to be as conservative as possible when it comes to the public API, because one of the biggest problems we can get long-term is to get locked in to a public API that we are not happy with because it prevents us from improving the library further.

This is obviously just my thoughts, and I am very open to being convinced otherwise. But I don't really see a need for clients to create tokens themselves. But you do, perhaps?

File: src/diff.rs
Comment (douweschulte) [reply src/diff.rs:22]: I do agree that it makes a more focussed discussion. But I needed to add the `ChangedToken` while working on the `Token` to make sure the chosen strategy made sense for for this use case as well. But if you want I will disentangle them. It depends if you are comfortable with such a large code change and feel that you can find all errors this way, or need it disentangled. I am willing to disentangle, but would like to know that my effort is 'needed' ;-).

**Comment by douweschulte** [comment]
Yes I do agree that we need to remove all token creation APII, I will scrutinize the API diff and remove all code related to that.

File: src/diff.rs
Comment (Enselic) [reply src/diff.rs:22]: Yeah, it totally get that it made a lot of sense to develop both concepts together to make sure things worked together. That was the right move. But now we that we have concepts in place that we know works, we are at a point where we can choose one of two scenarios:

Scenario A
1. Have tokens merged after 1 time units from now (PR 1)
2. Have intra-item diffing merged after 2 time units from now (PR 2)

Scenario B
1. Have tokens and intra-item diffing merged after 2 time units from now (one big PR)

Out of these scenarios I prefer A.

... but again, we do this purely for fun, so I would not say we *"need"* scenario A. and if you find it boring to work in an incremental manner (which I totally understand) we will not do that. So basically: your call :) But *my* preference would be to do a split.

Comment (douweschulte) [reply src/diff.rs:22]: Okay I will work on a split then. I do agree that it is more logical.

**Comment by douweschulte** [comment]
I `pub(crate)`ed everything that is not necessary outside this crate. I used the `TokenStream::extend` function in the `crate-public-api` to pad out the stream with whitespace. So I could also write a helper method for padding. Downstream users will still be able to crate tokens, but that is needed for nice matching so I would not worry about that. If wanted there could be a private phantomdata item added to `TokenStream` to keep downstream users from being able to create an instance.

Note the missing `PublicItem::ne` will fall back to the default implementation from the trait which is fine in this case.

<details>
<summary>cargo public-items --diff-git-checkouts upstream/main main</summary>

```
Removed items from the public API
=================================
-pub fn public_items::PublicItem::ne(&self, other: &PublicItem) -> bool

Changed items in the public API
===============================
- pub fn public_items::PublicItem::cmp(&self, other: &PublicItem) -> crate::cmp::Ordering
+                                       self          Self           std
- pub fn public_items::PublicItem::eq(&self, other: &PublicItem) -> bool
+                                      self          Self
- pub fn public_items::PublicItem::partial_cmp(&self, other: &PublicItem) -> crate::option::Option<crate::cmp::Ordering>
+                                               self          Self                                 std

Added items to the public API
=============================
+pub struct field public_items::PublicItem::tokens: TokenStream
+pub fn public_items::diff::ChangedPublicItem::clone(&self) -> ChangedPublicItem
+pub fn public_items::diff::PublicItemsDiff::clone(&self) -> PublicItemsDiff
+pub mod public_items::tokens
+pub enum public_items::tokens::Token
+pub enum variant public_items::tokens::Token::Function(String)
+pub enum variant public_items::tokens::Token::Generic(String)
+pub enum variant public_items::tokens::Token::Identifier(String)
+pub enum variant public_items::tokens::Token::Keyword(String)
+pub enum variant public_items::tokens::Token::Kind(String)
+pub enum variant public_items::tokens::Token::Lifetime(String)
+pub enum variant public_items::tokens::Token::Primitive(String)
+pub enum variant public_items::tokens::Token::Qualifier(String)
+pub enum variant public_items::tokens::Token::Self_(String)
+pub enum variant public_items::tokens::Token::Symbol(String)
+pub enum variant public_items::tokens::Token::Type(String)
+pub enum variant public_items::tokens::Token::Whitespace
+pub fn public_items::tokens::Token::clone(&self) -> Token
+pub fn public_items::tokens::Token::eq(&self, other: &Token) -> bool
+pub fn public_items::tokens::Token::fmt(&self, f: &mut crate::fmt::Formatter<'_>) -> crate::fmt::Result
+pub fn public_items::tokens::Token::len(&self) -> usize
+pub fn public_items::tokens::Token::ne(&self, other: &Token) -> bool
+pub fn public_items::tokens::Token::text(&self) -> &str
+pub struct public_items::tokens::TokenStream
+pub fn public_items::tokens::TokenStream::clone(&self) -> TokenStream
+pub fn public_items::tokens::TokenStream::default() -> TokenStream
+pub fn public_items::tokens::TokenStream::eq(&self, other: &TokenStream) -> bool
+pub fn public_items::tokens::TokenStream::extend(&mut self, tokens: impl Into<Self>)
+pub fn public_items::tokens::TokenStream::fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result
+pub fn public_items::tokens::TokenStream::fmt(&self, f: &mut crate::fmt::Formatter<'_>) -> crate::fmt::Result
+pub fn public_items::tokens::TokenStream::from(token: Token) -> Self
+pub fn public_items::tokens::TokenStream::from(tokens: Vec<Token>) -> Self
+pub fn public_items::tokens::TokenStream::from(tokens: &[Token]) -> Self
+pub fn public_items::tokens::TokenStream::is_empty(&self) -> bool
+pub fn public_items::tokens::TokenStream::len(&self) -> usize
+pub fn public_items::tokens::TokenStream::ne(&self, other: &TokenStream) -> bool
+pub fn public_items::tokens::TokenStream::tokens(&self) -> impl Iterator<Item = &Token> + '_
+pub struct field public_items::tokens::TokenStream::tokens: Vec<Token>
+pub fn public_items::tokens::TokenStream::tokens_len(&self) -> usize
+pub macro public_items::ws!
```
</details>

Besides that I created a single commit which removes all code related to token diffing. I am not really sure how well git will handle stuff like this if now both PRs are merged. If you know a better way to handle this please let me know.

**Comment by douweschulte** [comment]
All tests now pass except for the two that take in the output from a different file. These fail for some reason while the diff between both outputs (according to the test helper) is nothing.

File: README.md
Comment (Enselic) [inline README.md:7]: Not sure why the readme shows a diff here, maybe enough if you merge origin/main into your branch

File: src/tokens.rs
Comment (Enselic) [inline src/tokens.rs:21]: Currently, clients can create their own `Token`. If you put this test in lib_tests.rs it will compile:
```rs
#[test]
fn create_token() {
    let _: public_items::tokens::Token = public_items::tokens::Token::Function(String::from("banana"));
}
```

However, I would like to forbid clients to do that. Fortunately it is easy, just annotate each enum variant with `#[non_exhaustive]`:
```rs
    #[non_exhaustive]
    Function(String),
```

we should also annotate the enum itself as `#[non_exhaustive]`:
```rs
#[non_exhaustive]
pub enum Token {
```

**Review by Enselic — COMMENTED** [review]
I won't be able to go through it all at once, so here are a couple of early comments. I plan on leaving more comments later as I go along doing the review.

**Review by Enselic — COMMENTED** [review]
I think tests are failing a bit too much for me to do a very detailed review. Do you think you could make all tests pass before the next review round? Let me know if you think that would be too much work. Again, no stress, it is fine if it takes you a while.

I feel like maybe we can simplify render.rs a bit? But before all tests pass there is no point in trying to figure out how, because before tests pass we do not know under what constraints we can refactor.

File: src/tokens.rs
Comment (douweschulte) [reply src/tokens.rs:21]: Forbidding downstream users from creating `Token` instances makes pattern matching very hard.

With `#[non_exhaustive]` on the whole enum all downstream crates have to have a catch all arm `_`. Which pretty much cannot do anything else aside from panic or plain skip. I would argue that having the compiler warn you when you upgrade a dependency (as long as it is following semver) is perfectly fine, especially because adding an extra arm in this case should be pretty straightforward.

With `#[non_exhaustive]` on all items the downstream users need to match every item with extra fields (see below) which I think does not really add to the ease of use.
```rust
match token {
    Token::identifier(text, ..) => ()
}
```

While I do agree that in the general case having users create these items for themselves, I would argue that having nice pattern matching is more important. We should however make sure that users cannot brake this crate by creating their own `Token`s which is the case, because there is no way of feeding your own `Token`s back into the library.

File: src/tokens.rs
Comment (douweschulte) [reply src/tokens.rs:35]: I do agree, I was trying to figure out how and this works perfectly!

File: src/item_iterator.rs
Comment (douweschulte) [reply src/item_iterator.rs]: I reverted the sort order, I liked the new one because it sorted related items (same path) together so the new enum and its fields would be close while in the previous one it sorted item kinds (fn/struct/etc) together. But this can be very easily adding in later.

For the two examples. the first one I deliberately deleted the `$` because that is not what a user would see in their own IDE and replaced it with a `Token` that is different from the standard, so it retains the information.

For the second one I will dive into how to restore that, good catch!

**Comment by douweschulte** [comment]
I thought I got all tests to pass, but I missed all the nice ones. Because two were randomly crashing in `bin_lib` I never got to the nice ones in `lib_test`. I fixed all except some in `comprehensive_api` for which I need to do some more work on the generics rendering. For structs I render them as a comma separated list of all item names, do you agree to handle it this way?

And I do agree to keep the sort order out of this PR otherwise the output from the tests is just unusable.

<details>
<summary>The diff, with only the changed lines highlighted</summary>

_Preview looks terrible if you click to enlarge it looks fine_
image
</details>

File: README.md
Comment (douweschulte) [reply README.md]: Yes that helped.

File: src/tokens.rs
Comment (Enselic) [reply src/tokens.rs:21]: Ah right, I forgot about pattern matching. I agree that is too unergonomic. So let's not mark enum variants `#[non_exhaustive]`.

You are also right about that with `#[non_exhaustive]` on the whole enum, it is not really possible to do something sensible in the catch-all arm. We *maybe* could make tokens annotate text, rather than containing text, to make a catch-all arm doable, but I'm fine with if we don't do that now in this PR.

So in short, you don't need to add `#[non_exhaustive]` anywhere. Thank you for reminding me of the bigger picture.

**Comment by Enselic** [comment]
> * [ ]  For structs I render them as a comma separated list of all item names, do you agree to handle it this way?

I'm not a fan of that to be honest, because it will make diffing more noisy. In general I like diffs to be on an item-level. I am open to changing that later, but if it is OK with you I would like to not do it in this PR. Also of importance: `cargo doc` HTML does not display structs with their fields like that, and I consider it quite important that our output is the same as what cargo doc has, so that users feels familiar with the output.

> * [ ]  I deliberately deleted the `$` from `$crate` because that is not what a user would see in their own IDE and replaced it with a `Token` that is different from the standard, so it retains the information in the `TokenStream`.

Our "gold standard" is `cargo doc` HTML output, and it turns out they do not render `$crate`. So long term we should render it the same. Short term however, I'm not entirely comfortable to change it just slightly. I am open to discuss and maybe that later, but for this PR I think it is good if we keep output the same as before, unless it is too much work.

Apart from the above it looks like it's only generics that indeed needs some tweaking.

Great work so far! Looking forward to your next update!

File: tests/expected_output/thiserror-1.0.30.txt
Comment (Enselic) [inline tests/expected_output/thiserror-1.0.30.txt:2]: Sorry, I created a conflict with you here with https://github.com/Enselic/public_items/pull/68

I try to not create conflicts but forgot to double-check this one

Fortunately it should be easy for you to resolve; change from
```
pub proc macro comprehensive_api_proc_macro::SimpleDeriveMacro!
```
to
```
pub proc macro comprehensive_api_proc_macro::#[derive(SimpleDeriveMacro)]
```
in ./tests/expected_output/comprehensive_api_proc_macro.txt

File: src/tokens.rs
Comment (douweschulte) [reply src/tokens.rs:21]: No problem, that's why we need two eyes on every piece of code! 
I am not so sure about having the text outside of the enum, if the need comes up we need a more detailed discussion I think.

File: tests/bin_tests.rs
Comment (Enselic) [inline tests/bin_tests.rs:9]: Just a bit of repetition here: It would be great if you could avoiding changing "expected output" (here for example). Not only to avoid conflicts with https://github.com/Enselic/public_items/pull/70 (which I will wait with merging until this PR is merged though), but also because good practice in and of itself, when doing big changes.

File: tests/expected_output/thiserror-1.0.30.txt
Comment (Enselic) [reply tests/expected_output/thiserror-1.0.30.txt:2]: > (The reason I work towards removing all .json from the repo is that that makes it easier to rename the lib to public-api later.)

Just a note (also mentioned in a separate comment): I now feel done with a big restructuring on how tests are run. Mainly stopping to use public_items itself for tests, because that becomes a real pain when we will rename the lib to public-api. The PR is at https://github.com/Enselic/public_items/pull/70, but I will not merge it before this PR is merged, to avoid causing more troubles for you. Again, I am very open to resolving the thiserror-1.0.30.txt-conflict for you, if you wish. It is not fair for you to solve problems caused by me :)

**Comment by Enselic** [comment]
@douweschulte Hi! I developed a [commit](https://github.com/Enselic/public_items/commit/11776e38f67accf7d1f299c9e95751835ddbc487) that makes all tests pass and that takes care of my own code review comments. Is it OK if I push it to this PR and then merge this PR? We can then of course iterate on the code further, but it is much nicer to do that in small increments in small PRs rather than have one big PR open :)

(Note that the commit does a "revert" of changes to expected output. To diff expected output, it is better to diff against v0.8.0 than towards the parent of my commit.)

If you think I am "stealing" your work or does something wrong here in any other way, please let me know!

**Comment by douweschulte** [comment]
Nice work! I do agree that small PRs are better here once this beast has been merged. I think it would be fine to merge and gather up all small leads in separate PRs. I was planning on doing this work this weekend but the fact that you build it is a pleasant surprise. This huge beast of a PR was getting a bit hard to fully understand every detail of all the time. And I got a surprise load of work to do in other projects this week.

So I would propose to merge this, I will then gather all loose ends from my side and open issues/PRs respectively to continue our discussion there. As I said I got some work to do on other projects as well so I maybe will be a bit less active for a while, but I want to continue work here as well.

**Comment by Enselic** [comment]
@douweschulte Great! I will merge this PR then, and shortly after I will also merge https://github.com/Enselic/public_items/pull/70. Stay tuned :)

**Review by Enselic — APPROVED** [review]

**Pull request merged**
